text/0000-rejoin-slice.md

```diff
@@ -0,0 +1,182 @@
+specific shorter ones after getting them from the library.
+
+As a remedy, it's often possible to use indices and/or pointers, and retrieve
+a longer subslice from the original slice. However this might be fiddly and bug-prone.
+The API proposed by this RFC provides an obvious and easy-to-use way to
+rejoin slices. 
+
+The said indexing workaround becomes even harder when mutable slices are used -
+as long as the derived subslices exist, the original slice stays borrowed
+and it's hard for the user to retrieve a longer slice in the case they
+would need it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Sometimes we get, as a result of parsing, a bunch of substrings that we know to be
+split from the original string:
+
+```
+let mut values: Vec<&str> = util_lib::split_by_streak("aaaaaaabbbbbbbcccccccddddeeeeeeefffggggggggh");
+assert_eq!(&"gggggggg", values[values.len()-1]);
+assert_eq!(&"h", values[values.len()-2]);
+```
+
+In the case we want to join adjacent substrings back together, we can use the `rejoin` API:
+
+```
+let last_two = &values[values.len()-2].rejoin(&values[values.len()-1]);
+assert_eq!(&"ggggggggh", last_two);
+```
+
+Slices have a similar API too. Joining subslices is especially useful when a parser returns mutable slices:
+
+```
+let mut values: Vec<&mut [u32]> = util_lib::split_by_streak_mut(&mut [5, 5, 5, 5, 2, 2, 7, 7, 7, 7, 7, 3, 3, 3][..]);
+assert_eq!(&mut [7, 7, 7, 7, 7][..], values[values.len()-2]);
+assert_eq!(&mut [3, 3, 3][..], values[values.len()-1]);
+```
+
+Modifying the other subslices as-is, but joining the last two into a single subslice
+would be cumbersome to do without the `rejoin` operation - we would have to get rid of the
+mutable slices to be able to index the original slice, which defeats the purpose
+of the operation we just did with `split_by_streak_mut`!
+
+```
+let last_two = &values[values.len()-2].rejoin_mut(&values[values.len()-1]);
+assert_eq!(&mut [7, 7, 7, 7, 7, 3, 3, 3][..], last_two);
+```
+
+Fortunately `rejoin_mut` allows us to easily combine the last two subslices back together.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes adding the following APIs to the standard library:
+
+```
+impl<T> [T] {
+    fn rejoin<'r>(&'r self, other: &'r [T]) -> &'r [T] {
+        self.try_rejoin(other).expect("the input slices must be adjacent in memory")
+    }
+
+    fn rejoin_mut<'r>(&'r mut self, other: &'r mut [T]) -> &'r mut [T] {
+        self.try_rejoin_mut(other).expect("the input slices must be adjacent in memory")
+    }
+
+    fn try_rejoin<'r>(&'r self, other: &'r [T]) -> Option<&'r [T]> {
+        let self_len = self.len();
+        let self_end = self[self_len..].as_ptr();
+        if core::ptr::eq(self_end, other.as_ptr()) {
+            Some(unsafe { core::slice::from_raw_parts(self.as_ptr(), self.len() + other.len()) })
```

Comment:
This is the unsoundness: `from_raw_parts` requires the entire range to be within a single allocation, which your check does not ensure (the two allocations could be situated right next to each other).

We should probably clarify the docs.